model/host/host.go

```diff
@@ -957,8 +975,74 @@ func (h *Host) setAwaitingJasperRestart(user string) error {
 	return nil
 }
 
-// MarkAsReprovisioning puts the host in a state that means it is going to be
-// reprovisioned.
+// SetNeedsReprovisionToNew marks a host that is currently using new
+// provisioning to provision again.
+func (h *Host) SetNeedsReprovisionToNew(user string) error {
+	// Ignore hosts that are not provisioned by us (e.g. hosts spawned by
+	// tasks) and legacy SSH hosts.
+	if utility.StringSliceContains([]string{distro.BootstrapMethodNone, distro.BootstrapMethodLegacySSH}, h.Distro.BootstrapSettings.Method) {
+		return nil
+	}
+
+	if err := h.setAwaitingReprovisionToNew(user); err != nil {
+		return err
+	}
+	if h.StartedBy == evergreen.User && !h.NeedsNewAgentMonitor {
```

Comment:
This only applies to hosts running agents and we don't have a direct check for this other than checking the user that started the host.

Reply:
There's two parts to reprovisioning - setting the correct DB state to ensure the provisioning job will run _eventually_ (with `seAwaitingReprovisionToNew()`) and ensuring the host has the correct DB state that it can run right now (`MarkAsReprovisioning()`). Before this line, `setAwaitingReprovisionToNew()` will just set the `NeedsReprovision` field to ensure the host has to reprovision at some point. However, since the agent might still be up and running a task, we can't set it in the state that would start the reprovisioning immediately (what `MarkAsReprovisioning()` does). The check for `NeedsNewAgentMonitor` verifies that, if the agent is not up currently, put it in the state that it can run the reprovision job right now.